### ISSUE TITLE
Change multicast code for Android

### ISSUE THREAD

**Issue by Revertron** [yggdrasil-network/yggdrasil-go#1149]
For fixing multicast in Android app (https://github.com/yggdrasil-network/yggdrasil-android/issues/60, https://github.com/yggdrasil-network/yggdrasil-android/issues/45) we need to change the way `multicast` module gets interface addresses.

One way for this is using `anet` library: https://github.com/wlynxg/anet

But I'm not a go developer... And we need to discuss what way is better.

**Comment by neilalexander** [comment]
Is this the one where Android prevents us from calling `net.Interfaces()` properly? If that's the case, we can just modify the `multicast` module to take a list of interfaces/addresses as input and then have Kotlin pass them in.

**Comment by Revertron** [comment]
Yes, it's the one! :)

> modify the multicast module to take a list of interfaces/addresses as input and then have Kotlin pass them in

Yep, I see this way as better. But how can we implement this? I mean which type of objects will Go part hold|get from Kotlin? Do we need to encode names+IPs in some intermediate format?

**Comment by bobrofon** [comment]
I tried to use [wlynxg/anet](https://github.com/wlynxg/anet) instead of the standard `net` package. `anet` is a little quirky library, but it solves at least part of the problem with multicast.

Here is the patch https://github.com/bobrofon/yggdrasil-go/commit/87234ea2de42512bccf8daaf248d42ed628ced85.
With this patch the multicast module is able to find multicast interfaces and even get the ip addresses. However local peer discovery still doesn't work.
```
I  Initializing multicast 
I  Starting multicast 
I  Starting multicast module
I  Started multicast module
I  Discovered addresses for interface wlan0: [192.168.1.204/24 fd76:dc0f:c24f:0:412b:6731:92c9:ccbf/64 fd76:dc0f:c24f:0:d7a4:c742:869e:f87d/64 fe80::8cf2:ff0:e65:b98e/64]
W  type=1400 audit(0.0:24825): avc: denied { read } for name="somaxconn" dev="proc" ino=1658926 scontext=u:r:untrusted_app:s0:c224,c256,c512,c768 tcontext=u:object_r:proc_net:s0 tclass=file permissive=0 app=eu.neilalexander.yggdrasil
I  Started multicasting on wlan0
I  TLS listener started on [fe80::8cf2:ff0:e65:b98e%wlan0]:37577
W  type=1400 audit(0.0:24826): avc: denied { bind } for scontext=u:r:untrusted_app:s0:c224,c256,c512,c768 tcontext=u:r:untrusted_app:s0:c224,c256,c512,c768 tclass=netlink_route_socket permissive=0 bug=b/155595000 app=eu.neilalexander.yggdrasil
W  type=1400 audit(0.0:24827): avc: denied { bind } for scontext=u:r:untrusted_app:s0:c224,c256,c512,c768 tcontext=u:r:untrusted_app:s0:c224,c256,c512,c768 tclass=netlink_route_socket permissive=0 bug=b/155595000 app=eu.neilalexander.yggdrasil
W  type=1400 audit(0.0:24828): avc: denied { bind } for scontext=u:r:untrusted_app:s0:c224,c256,c512,c768 tcontext=u:r:untrusted_app:s0:c224,c256,c512,c768 tclass=netlink_route_socket permissive=0 bug=b/155595000 app=eu.neilalexander.yggdrasil
```
In `strace` I can see that it receives some packages from the other node in the same network, though.
```
[pid  9223] recvmsg(61, {msg_name={sa_family=AF_INET6, sin6_port=htons(9001), inet_pton(AF_INET6, "fe80::345d:2708:6654:54f4", &sin6_addr), sin6_flowinfo=htonl(0), sin6_scope_id=if_nametoindex("wlan0")}, msg_namelen=112->28, msg_iov=[{iov_base="\0\0\0\5\271o\316\333\212\367h,\234\262j)k\220\356\30;5\303\306\256\316\24\267'\347X\25"..., iov_len=2048}], msg_iovlen=1, msg_control=[{cmsg_len=36, cmsg_level=SOL_IPV6, cmsg_type=0x32}], msg_controllen=40, msg_flags=0}, 0) = 104
[pid  9223] getsockopt(61, SOL_SOCKET, SO_PROTOCOL, [17], [4]) = 0
[pid  9223] socket(AF_NETLINK, SOCK_RAW|SOCK_CLOEXEC, NETLINK_ROUTE) = 78
[pid  9223] bind(78, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 12) = -1 EACCES (Permission denied)
[pid  9223] close(78 <unfinished ...>
```

**Comment by bobrofon** [comment]
It looks like [ipv6.PacketConn.ReadFrom](https://pkg.go.dev/golang.org/x/net/ipv6#PacketConn.ReadFrom) and [ipv6.PacketConn.WriteTo](https://pkg.go.dev/golang.org/x/net/ipv6#PacketConn.WriteTo) implementations internally use `net.Interfaces`. This means that they cannot be used in the multicast module on Android.
```
#0  runtime/internal/syscall.Syscall6 () at /usr/lib/golang/src/runtime/internal/syscall/asm_linux_arm64.s:17
#1  0x00000073461efb8c in syscall.RawSyscall6 (num=18446744073709551603, a1=274880308844, a2=12, a3=0, a4=0, a5=0, a6=0, r1=<optimized out>, 
    r2=<optimized out>, errno=<optimized out>) at /usr/lib/golang/src/runtime/internal/syscall/syscall_linux.go:38
#2  0x00000073462734d0 in syscall.Syscall (trap=200, a1=119, a2=274880308844, a3=12, r1=<optimized out>, r2=<optimized out>, err=<optimized out>)
    at /usr/lib/golang/src/syscall/syscall_linux.go:82
#3  0x0000007346272980 in syscall.bind (s=<optimized out>, addr=<optimized out>, addrlen=<optimized out>, err=...)
    at /usr/lib/golang/src/syscall/zsyscall_linux_arm64.go:1392
#4  0x0000007346270678 in syscall.Bind (fd=119, sa=..., err=...) at /usr/lib/golang/src/syscall/syscall_unix.go:290
#5  0x000000734626ea3c in syscall.NetlinkRIB (proto=18, family=0, ~r0=..., ~r0=..., ~r1=..., ~r1=...) at /usr/lib/golang/src/syscall/netlink_linux.go:67
#6  0x000000734634ffe0 in net.interfaceTable (ifindex=0, ~r0=..., ~r1=...) at /usr/lib/golang/src/net/interface_linux.go:17
#7  0x000000734634f1c0 in net.Interfaces (~r0=..., ~r1=...) at /usr/lib/golang/src/net/interface.go:103
#8  0x000000734651a87c in golang.org/x/net/internal/socket.(*ipv6ZoneCache).update (zc=0x7346989960 <golang.org/x/net/internal/socket[zoneCache]>, ift=..., 
    force=true, updated=<optimized out>, updated=<optimized out>) at /home/user/go/pkg/mod/golang.org/x/net@v0.27.0/internal/socket/sys_posix.go:137
#9  0x000000734651ac8c in golang.org/x/net/internal/socket.(*ipv6ZoneCache).name (zc=<optimized out>, zone=32, ~r0=...)
    at /home/user/go/pkg/mod/golang.org/x/net@v0.27.0/internal/socket/sys_posix.go:158
#10 0x000000734651b6e0 in golang.org/x/net/internal/socket.sockaddrToAddr (sa=..., network=..., ~r0=...)
    at /home/user/go/pkg/mod/golang.org/x/net@v0.27.0/internal/socket/sys_unix.go:107
#11 0x000000734651b0c4 in golang.org/x/net/internal/socket.recvmsg (s=<optimized out>, buffers=..., oob=..., flags=<optimized out>, network=..., 
    n=<optimized out>, oobn=<optimized out>, recvflags=<optimized out>, from=..., err=...)
    at /home/user/go/pkg/mod/golang.org/x/net@v0.27.0/internal/socket/sys_unix.go:36
#12 0x0000007346519c58 in golang.org/x/net/internal/socket.(*Conn).recvMsg.func1 (s=18446744073709551603, ~r0=<optimized out>)
    at /home/user/go/pkg/mod/golang.org/x/net@v0.27.0/internal/socket/rawconn_msg.go:24
#13 0x00000073462c0b48 in internal/poll.(*FD).RawRead (fd=0x4000117680, f={void (uintptr, bool *)} 0x40003d1aa0, ~r0=..., ~r0=...)
    at /usr/lib/golang/src/internal/poll/fd_unix.go:705
#14 0x000000734635c900 in net.(*rawConn).Read (c=0x400013a0c0, f={void (uintptr, bool *)} 0x0, ~r0=...) at /usr/lib/golang/src/net/rawconn.go:44
#15 0x0000007346519ab8 in golang.org/x/net/internal/socket.(*Conn).recvMsg (c=0x40001430e0, m=0x400020e3c0, flags=0, ~r0=...)
    at /home/user/go/pkg/mod/golang.org/x/net@v0.27.0/internal/socket/rawconn_msg.go:27
#16 0x0000007346520124 in golang.org/x/net/internal/socket.(*Conn).RecvMsg (c=0xfffffffffffffff3, m=0x400020e3c0, flags=<optimized out>)
    at /home/user/go/pkg/mod/golang.org/x/net@v0.27.0/internal/socket/socket.go:247
#17 golang.org/x/net/ipv6.(*payloadHandler).ReadFrom (c=0x40003c6920, b=..., n=<optimized out>, cm=<optimized out>, src=..., err=...)
    at /home/user/go/pkg/mod/golang.org/x/net@v0.27.0/ipv6/payload_cmsg.go:31
#18 0x0000007346586e50 in github.com/yggdrasil-network/yggdrasil-go/src/multicast.(*Multicast).listen (m=0x40002541c0)
    at /home/user/yggdrasil/yggdrasil-go/src/multicast/multicast.go:445
#19 0x00000073465847b8 in github.com/yggdrasil-network/yggdrasil-go/src/multicast.(*Multicast)._start.gowrap3 ()
    at /home/user/yggdrasil/yggdrasil-go/src/multicast/multicast.go:119
#20 0x0000007346263684 in runtime.goexit () at /usr/lib/golang/src/runtime/asm_arm64.s:1222
```

**Comment by neilalexander** [comment]
It might just be that the right answer here is to implement the multicast code natively in Kotlin and not use the Go multicast code on Android, if Android is going to be this awkward about it. 

Android is the only platform we seem to have these issues on.

**Comment by bobrofon** [comment]
As far as I understand, `anet` package didn't work properly with `yggdrasil-go` because `yggdrasil-go` uses the `golang.org/x/net` package instead of the `net` package from the standard library. I tried to add `golang.org/x/net` support to the `anet` package, and now multicasting works as expected.
Here is the final patch I used: https://github.com/bobrofon/yggdrasil-go/commit/95ce60abe502934c998b9e2386daa1f174219b67.

> It might just be that the right answer here is to implement the multicast code natively in Kotlin

I agree, but someone needs to write this implementation first. And `anet' library looked like an easy and fast workaround. It was worth trying, I think.

By the way, I read on some wiki pages about how some people use `yggdrasil-go` on Android from command line (without `yggdrasil-android`) using `termux`. So keeping working Go multicast implementation on Android may have some benefits too, I guess.
